In [ ]:
import pandas as pd
import numpy as np

old_data = pd.read_csv("music_old.csv")
new_data = pd.read_csv("music_new.csv")

col = "loudness"
n_bins = 20

quantiles = np.linspace(0, 1, n_bins + 1)
bin_edges = old_data[col].quantile(quantiles).values

old_counts, _ = np.histogram(old_data[col], bins=bin_edges)
old_percentiles = old_counts / old_counts.sum()

new_counts, _ = np.histogram(new_data[col], bins=bin_edges)
new_percentiles = new_counts / new_counts.sum()

epsilon = 1e-8
old_percentiles = np.clip(old_percentiles, epsilon, 1)
new_percentiles = np.clip(new_percentiles, epsilon, 1)

psi_values = (old_percentiles - new_percentiles) * np.log(old_percentiles / new_percentiles)
psi = np.sum(psi_values)

print(f"Population Stability Index (PSI) for column '{col}': {psi:.6f}") 

Population Stability Index (PSI) for column 'loudness': 0.235367


Выводы:

PSI для признака loudness составил ≈ 0.235, что превышает порог 0.2 и указывает на существенное смещение распределения новых данных относительно эталонных. Это означает, что громкость треков в новой выборке распределена иначе, чем в данных, на которых обучалась модель, — модель может работать менее точно на новых данных. Следует выяснить причину сдвига (например, изменение стандартов мастеринга или состав каталога) и рассмотреть переобучение модели на актуальных данных.